Add support for adau7118

nunojsa opened the pull request:
This patch series adds support for the 8 channel PDM-to-I2S/TDM converter. The ADAU7118 converts four stereo pulse density modulation (PDM) bitstreams into one pulse code modulation (PCM) output stream. The source for the PDM data can be eight microphones or other PDM sources. The PCM audio data is output on a serial audio interface port in either inter-IC serial (I2S) or time domain multiplexed (TDM) format.

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+	SOC_DAPM_SINGLE("Capture Switch", ADAU7118_REG_ENABLES, 3, 1, 0),
+};
+
+
```
nitpick: extra line

nunojsa replied:
ack

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+
+	switch (fmt & SND_SOC_DAIFMT_FORMAT_MASK) {
+	case SND_SOC_DAIFMT_I2S:
+		/* reset value */
```
curios: should there be something here?
the comment is also a bit vague; maybe drop it or expand it?

commodo replied:
i think i gt the gist of it now;
question is: do you need to do something more here? i mean, i guess `SND_SOC_DAIFMT_I2S` is the value on reset, but can this be changed to SND_SOC_DAIFMT_LEFT_J and then back to SND_SOC_DAIFMT_I2S during runtime?
and if yes, do you need to update some bits?

nunojsa replied:
Hmm, that's a nice question. I don't think we can change it in runtime. The fmt is a devicetree configuration on the sound card. However, I'm not 100% sure on this so I guess I will also update the value in this case to be on the safe side.

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+	struct adau7118_data *st = snd_soc_component_get_drvdata(component);
+	int ret = 0;
+
+	dev_info(st->dev, "Set bias level %d\n", level);
```
should this be dev_dbg?

nunojsa replied:
Definitely. Thanks

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+
+			/* there's no timing constraints before enabling dvdd */
+			ret = regulator_enable(st->dvdd);
+			if (ret)
```
should `st->iovdd` be disabled on this error path?

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+		if (!st->iovdd)
+			return 0;
+
+		ret = regulator_disable(st->dvdd);
```
with a bit of rework, adau7118_regulator_disable() could be used here

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+
+		ret = regulator_disable(st->dvdd);
+		if (ret)
+			return ret;
```
also, not sure if it's a good idea to return here ; `st->iovdd` won't be disabled

nunojsa replied:
That's on purpose. If we fail to disable dvdd, we don't want to disable iovdd. The part should not have dvdd applied and not iovdd. In fact, the only constraint the datasheet stats on the power up is that, dvdd cannot be applied before iovdd.

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+	if (!ret) {
+		switch (dec_ratio) {
+		case 64:
+			/* reset value */
```
i am thinking that, even if this is a reset-value, it may be a good idea to write it [if possible];
there are some cases, like warm-reboot, where the system may not go through some power-resets, and it would be a good idea to update reset-values

nunojsa replied:
ack

nunojsa replied:
There is a RESETS register where we can perform a full SW reset on the device (including register settings): I wonder if it does not make sense to do this at `probe()`?. Exactly by the same reason as you stated (warm-reboot).

commodo commented on sound/soc/codecs/adau7118.h:
```diff
@@ -0,0 +1,24 @@
+#define ADAU7118_REG_DRIVE_STRENGTH	0x11
+#define ADAU7118_REG_RESET		0x12
+
+int adau7118_probe(struct device *dev, struct regmap *map, bool hw_mode);
```
does this need an `extern` for when this is kmod built?

nunojsa replied:
I don't think so. But `adau7118_probe()` definitely needs to be exported with `EXPORT_SYMBOL_GPL`.

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,619 @@
+	return devm_snd_soc_register_component(dev,
+					       &adau7118_component_driver,
+					       &adau7118_dai, 1);
+}
```
does this require some `MODULE_` macros? for kmod build?

nunojsa replied:
Don't think so. The macros are defined on the drivers which are `adau7118-i2c` and `adau7118-hw`

nunojsa replied:
Forget it. It definitely needs the macros

commodo commented on sound/soc/codecs/adau7118-hw.c:
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: GPL-2.0
```
license-wise [when upstreaming] i think this may need to be `GPL-2.0+` or, `MODULE_LICENSE("GPL v2");`
i am not sure about these [and which is correct], but some upstream people really care about this subtle-legal-stuff;

nunojsa replied:
Also not sure, I think I already saw some points here (on some other reviews) that we should go with `MODULE_LICENSE("GPL");`. So I try to be consistent and always use this form.

commodo commented on sound/soc/codecs/Kconfig:
```diff
@@ -374,6 +376,32 @@ config SND_SOC_ADAU1977_I2C
+config SND_SOC_ADAU7118
+	tristate
+
+config SND_SOC_ADAU7118_HW
```
the HW mode looks a bit weird;
it makes the role of a driver [generally] useless;
if it just needs to control some regulators, then maybe some other "fixed" regulator driver would be an idea or something?

nunojsa replied:
Not only that. Still you do need to register the device on the asoc subsystem and provide the right routes + widgets. Look at https://github.com/analogdevicesinc/linux/blob/master/sound/soc/codecs/adau7002.c as an example.

commodo replied:
ack
thanks for the clarification

nunojsa commented:
Changes in V2:

- Always update device configurations (even if it's the reset value);

- Removed blank line;

- Converted `dev_info` into `dev_dbg`;

- Added `MODULE_` macros to adau7118.c;

- Added `EXPORT_SYMBOL_GPL(adau7118_probe)`;

- Disable `st->iovdd` if we fail to enable `st->dvdd`;

- Perform full SW reset on `probe()`;

- Corrected the name of the drivers (when compiled as modules) in Kconfig;

- Removed `adi,decimation-ratio`property from HW mode example in the yaml binding;

- Added phandle `$ref` to regulator properties.

commodo commented on sound/soc/codecs/adau7118.c:
```diff
@@ -0,0 +1,645 @@
+		ret = snd_soc_component_update_bits(dai->component,
+						ADAU7118_REG_SPT_CTRL2,
+						ADAU7118_LRCLK_BCLK_POL_MASK,
+						ADAU7118_LRCLK_BCLK_POL(0));
```
it looks like in switch statements, things could be simplified a bit:
```
	case SND_SOC_DAIFMT_NB_NF:
		ret = snd_soc_component_update_bits(dai->component,
						ADAU7118_REG_SPT_CTRL2,
						ADAU7118_LRCLK_BCLK_POL_MASK,
						ADAU7118_LRCLK_BCLK_POL(0));
```

can become:
```
	case SND_SOC_DAIFMT_NB_NF:
		regval = ADAU7118_LRCLK_BCLK_POL(0);
		break;

..................
		ret = snd_soc_component_update_bits(dai->component,
						ADAU7118_REG_SPT_CTRL2,
						ADAU7118_LRCLK_BCLK_POL_MASK,
						regval);
```

or even just a lookup table maybe / depends on how many switch() statements could be reduced;

nunojsa replied:
Yes, I guess I could that. But IMHO, as is, it is pretty readable and explicit what is being done. I'm not seeing any big advantage with the simplification as the switch() should not grow bigger that what it is now. I would not go for the lookup table. As for the simplification, I don't feel to strong about it so I can do it if there's any other reason for a v3. But I would say to leave as is, if nothing else needs to be changed (so no v3 is needed).

mhennerich replied:
Everything that reduces the SLOC is worth doing it...

nunojsa replied:
Honestly, I'm still not 100% convinced but I did found the need for a v3 so I will make the change.

nunojsa commented:
Changes in V3:

- Corrected some error paths taking into account that `snd_soc_component_update_bits()` returns 1 if the operation was successful and the value of the register changed;
- Refactored `switch() ... cases:` where appropriated;

commodo commented:
looks good to me;

shall we close this PR and send it upstream?

commodo commented:
new idea: let's cherry-pick these patches to `master-xilinx-4.19` `adi-4.19.0` & `rpi-4.19.y`

thoughts? objections?

nunojsa commented:
No objections from my side...

commodo commented:
Closing.

Commits have been manually added to branches `master-xilinx-4.19` , adi-4.19.0` & `rpi-4.19.y`

Their hashes are 21f264f6c65de1d5e2a6440a8012dbca121e6e37  &  08d40b796e96a629f01a8b8868998c27ee539ec9  for branch master-xilinx-4.19; which are the base-official commits for this driver in the ADI Linux repo.

And for the DT overlay in rpi-4.19.y is e478ee6a302453e8953182f0cf9c9fbc7714ac81

Branch `master-xilinx-4.19`  will be merged to `master` at some point, so the original driver hashes in `master-xilinx-4.19` should be preserved.
